core/src/HIP/Kokkos_HIP_Instance.cpp

```diff
@@ -461,42 +408,61 @@ HIP::size_type HIP::device_arch()
 }
 #endif
 
-void HIP::finalize()
+void HIP::impl_finalize()
 {
-  Kokkos::Impl::HIPInternal::singleton().finalize();
+  Impl::HIPInternal::singleton().finalize();
 
   #if defined(KOKKOS_ENABLE_PROFILING)
     Kokkos::Profiling::finalize();
   #endif
 }
 
 HIP::HIP()
-  : m_device( Kokkos::Impl::HIPInternal::singleton().m_hipDev )
+  : m_space_instance(&Impl::HIPInternal::singleton())
 {
-  Kokkos::Impl::HIPInternal::singleton().verify_is_initialized( "HIP instance constructor" );
+  Impl::HIPInternal::singleton().verify_is_initialized( "HIP instance constructor" );
 }
 
 //HIP::HIP( const int instance_id )
 //  : m_device( Impl::HIPInternal::singleton().m_hipDev )
 //{}
 
 void HIP::print_configuration( std::ostream & s , const bool )
-{ Kokkos::Impl::HIPInternal::singleton().print_configuration( s ); }
+{ Impl::HIPInternal::singleton().print_configuration( s ); }
 
 bool HIP::sleep() { return false ; }
 
 bool HIP::wake() { return true ; }
 
 void HIP::fence()
 {
-  hipDeviceSynchronize();
+  HIP_SAFE_CALL( hipDeviceSynchronize() );
 }
 
+int HIP::hip_device() const { return impl_internal_space_instance()->m_hipDev; }
 const char* HIP::name() { return "HIP"; }
 
 } // namespace Experimental
+
+namespace Impl {
+void hip_device_synchronize()
+{
+  HIP_SAFE_CALL( hipDeviceSynchronize() );
+}
+
+void hip_internal_error_throw( hipError_t e , const char * name, const char * file, const int line )
+{
+  std::ostringstream out ;
+  out << name << " error( " << hipGetErrorName(e) << "): " << hipGetErrorString(e);
```

Comment:
No we shouldn't wrap all. HIP_SAFE_CALL will add a few micro seconds latency (since it involves a driver query to the GPU) so we should only add it where it doesn't hurt.